tsl/src/nodes/decompress_chunk/planner.c

```diff
@@ -454,29 +485,122 @@ decompress_chunk_plan_create(PlannerInfo *root, RelOptInfo *rel, CustomPath *pat
+		/* batch_merge_append is used when the 'order by' of the query and the
+		 * 'order by' of the segments do match, we use a heap to merge the segments.
+		 * For the heap we need a compare function that determines the heap order. This
+		 * function is constructed here.
+		 */
+		AttrNumber *sortColIdx = NULL;
+		Oid *sortOperators = NULL;
+		Oid *collations = NULL;
+		bool *nullsFirst = NULL;
+		int numsortkeys = 0;
+
+		ts_prepare_sort_from_pathkeys(&decompress_plan->scan.plan,
+									  dcpath->cpath.path.pathkeys,
+									  bms_make_singleton(dcpath->info->chunk_rel->relid),
+									  NULL,
+									  false,
+									  &numsortkeys,
+									  &sortColIdx,
+									  &sortOperators,
+									  &collations,
+									  &nullsFirst);
+
+		List *sort_col_idx = NIL;
+		List *sort_ops = NIL;
+		List *sort_collations = NIL;
+		List *sort_nulls = NIL;
+
+		/* Since we have to keep the sort info in custom_private, we store the information
+		 * in copyable lists */
+		for (int i = 0; i < numsortkeys; i++)
+		{
+			sort_col_idx = lappend_oid(sort_col_idx, sortColIdx[i]);
+			sort_ops = lappend_oid(sort_ops, sortOperators[i]);
+			sort_collations = lappend_oid(sort_collations, collations[i]);
+			sort_nulls = lappend_oid(sort_nulls, nullsFirst[i]);
+		}
+
+		sort_options = list_make4(sort_col_idx, sort_ops, sort_collations, sort_nulls);
+
+		/* Build a sort node for the compressed batches. The sort function is derived from the sort
+		 * function of the pathkeys, except that it refers to the min and max elements of the
+		 * batches. We have already verified that the pathkeys match the compression order_by, so
+		 * this mapping can be done here. */
+		for (int i = 0; i < numsortkeys; i++)
+		{
+			Oid opfamily, opcintype;
+			int16 strategy;
+
+			/* Find the operator in pg_amop --- failure shouldn't happen */
+			if (!get_ordering_op_properties(sortOperators[i], &opfamily, &opcintype, &strategy))
+				elog(ERROR, "operator %u is not a valid ordering operator", sortOperators[i]);
+
+			Assert(strategy == BTLessStrategyNumber || strategy == BTGreaterStrategyNumber);
+			char *meta_col_name = strategy == BTLessStrategyNumber ?
+									  column_segment_min_name(i + 1) :
+									  column_segment_max_name(i + 1);
+
+			AttrNumber attr_position =
+				get_attnum(dcpath->info->compressed_rte->relid, meta_col_name);
+
+			if (attr_position == InvalidAttrNumber)
+				elog(ERROR, "couldn't find metadata column \"%s\"", meta_col_name);
+
+			/* If the the target list is not based on the layout of the uncompressed chunk,
+			 * (see comment for physical_tlist above), adjust the position of the
+			 */
+			if (target_list_compressed_is_physical)
+				sortColIdx[i] = attr_position;
+			else
+				sortColIdx[i] =
+					find_attr_pos_in_tlist(compressed_scan->plan.targetlist, attr_position);
+		}
+
+		/* Now build the compressed batches sort node */
+		Sort *sort = ts_make_sort((Plan *) compressed_scan,
```

Comment:
I think instead of building the `Sort` here - you could more easily add it during the `Path` level planning phase - in that case things will be more straight

Reply:
Could you elaborate on this idea? 

We already do something similar for other paths and we should be consistent in the construction of the paths:

https://github.com/timescale/timescaledb/blob/d3730a4f6ac89612d749d10843def181f7830635/tsl/src/nodes/decompress_chunk/planner.c#L463

Reply:
I know - and I think that should be removed that as well!
I suspect that somehow that line causes this issue: https://github.com/timescale/timescaledb/issues/5507

Are there any benefits of creating the `Sort` node by hand and not instructing the system to create one via a `Path`?

Reply:
Creating this sort is a bit more effort (since it's not directly based on the path keys and we're doing attribute mapping). I want to delay this planning effort and only do it when we really create this plan.

Reply:
If you look at the merge append implementation in postgres you can see postgres handles this similarly. That is adding the cost for the sort during path creation but actually creating the sort node only during plan creation.

Reply:
I've seen that - but I still don't understand why should we follow that "similarity"? what are the benefits?

the code you are mentioning was added by [this commit](
https://github.com/postgres/postgres/commit/959d00e9dbe4cfcf4a63bb655ac2c29a5e579246) 
which seemed like a copy paste of the mergeappend path around that time.

but ever since then some things changed around mergeappend and there was a [commit](https://github.com/postgres/postgres/commit/6bc27698324a3eadb8a4a5a79d96e16687be6141) 

I think especially the commit message of this last one is important...
```
Ensuring this is the responsibility of the code constructing the paths,
```

I think the other one have not caused any issues so far  - so it was left as-is...

Reply:
Hmm that sounds like it is based on speculation from your side. GatherMerge and MergeAppend have a different purpose and appear in different parts of the plan. In a non-parallel plan there will never be a GatherMerge. That commit is specifically about GatherMerge and if you read up on the discussion around the commit you can see that the context of the change is regarding parallel plans.

Reply:
sure; I can be wrong - but I also think that `this is the responsibility of the code constructing the paths`...that's how things supposed to work...of course I accept valid reasons for not doing so...

...but I'm still struggling to understand why should we follow that - an do things similarily? what are the benefits of doing so? Could you answer this?

Reply:
In general i think is a good approach to follow the way postgres implements things. We are doing that in a lot of places inside timescaledb. In the case of merge append i think the benefit is in delaying creating additional planner nodes for pathes that might not be needed, reducing planning time. Why should we do things different from postgres when we are a postgres extension?

Reply:
interesting - I'm not sure I see the benefits in this case....

* note that postgres creates an >ordered< append ; which has multiple inputs (`k`); the counterargument that it was not sorted during planning phase could be:
  * I think normal append takes no gurantees about the order in which it gives back results
  * it might have not entirely clear it could be an ordered append during path creation
  * but it is true for sure that: even in the worst case running `n*log(n/k)` sorts is always better than doing `n*log(n)`
    * speedup is `log(1/k)` ; with better locality
  * so I think pushing in the sort in that case have its valid reasons
* in this; we have a `DC` node and beneath a single `input`....I think its a completely different situation
* I think the actual benefits you are mentioning in planning time may be imaginary - but we do pay with real added complexity for this unknown benefit

but lets put these aside - do you think if someone creates an index on the underlying table which could make this sort unneccesary - it will be easy to built that into this solution easily?
because at path level planning you could handle that easily...